shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java

```diff
@@ -75,14 +88,30 @@ public void send(
 
   @Override
   public void handleMessageFromDart(
-      @NonNull final String channel, @Nullable byte[] message, final int replyId) {
+      @NonNull final String channel, @Nullable ByteBuffer message, final int replyId) {
     Log.v(TAG, "Received message from Dart over channel '" + channel + "'");
-    BinaryMessenger.BinaryMessageHandler handler = messageHandlers.get(channel);
+    Handler handler = messageHandlers.get(channel);
     if (handler != null) {
       try {
         Log.v(TAG, "Deferring to registered handler to process message.");
-        final ByteBuffer buffer = (message == null ? null : ByteBuffer.wrap(message));
-        handler.onMessage(buffer, new Reply(flutterJNI, replyId));
+        ByteBuffer messageWithDirectness;
+        if (handler.wantsDirectByteBufferForDecoding) {
+          messageWithDirectness = message;
+        } else if (message != null) {
+          ByteBuffer indirectByteBuffer = ByteBuffer.allocate(message.capacity());
+          indirectByteBuffer.put(message);
+          indirectByteBuffer.rewind();
+          messageWithDirectness = indirectByteBuffer;
+        } else {
+          messageWithDirectness = null;
```

Comment:
It's not redundant because you'll get a static error that you are using messageWithDirectness when it is potentially not initialized.  I can't remember if it is the compiler or our linter that generates it, but this gives us static analysis that all branches set `messageWithDirectness`.